### PR TITLE
Merge 2024-marketplace into master

### PR DESCRIPTION
**Pull request by aembler**

### CODE REVIEW

**Review by wordish — APPROVED** [review]

File: src/PortlandLabs/ConcreteCmsTheme/User/CommunityUser.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace PortlandLabs\ConcreteCmsTheme\User;
+
+use Concrete\Core\Entity\User\User;
+use Concrete\Core\User\Avatar\AvatarInterface;
+use PortlandLabs\ConcreteCmsTheme\User\EmptyAvatar;
+use League\Url\Url;
+
+class CommunityUser
```
Comment (KorvinSzanto) [inline src/PortlandLabs/ConcreteCmsTheme/User/CommunityUser.php:10]: I didn't see this class get used anywhere, is it used in such a way that `CommunityUserInspector` can't just be a dependency?

File: src/PortlandLabs/ConcreteCmsTheme/User/CommunityUserAvatar.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        protected string $username,
+    ) {
+    }
+
+    public function getPath()
+    {
+        return $this->path;
+    }
+
+    public function output()
```
Comment (KorvinSzanto) [inline src/PortlandLabs/ConcreteCmsTheme/User/CommunityUserAvatar.php:26]:
```suggestion
    public function output(): string
```

File: src/PortlandLabs/ConcreteCmsTheme/User/CommunityUserAvatar.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+class CommunityUserAvatar implements AvatarInterface
+{
+
+    public function __construct(
+        protected string $path,
+        protected string $username,
+    ) {
+    }
+
+    public function getPath()
```
Comment (KorvinSzanto) [inline src/PortlandLabs/ConcreteCmsTheme/User/CommunityUserAvatar.php:21]:
```suggestion
    public function getPath(): string
```

File: src/PortlandLabs/ConcreteCmsTheme/User/CommunityUserInspector.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    public function __construct(
+        protected Connection $db,
+        protected ClientBuilder $clientBuilder,
+        protected ExpensiveCache $cache,
+    ) {
+    }
+
+    public function getCommunityUserData(User $user): ?array
+    {
+        $cacheKey = '/community/user/' . $user->getUserID();
```
Comment (KorvinSzanto) [inline src/PortlandLabs/ConcreteCmsTheme/User/CommunityUserInspector.php:22]:
```suggestion
        $cacheKey = 'community.user.' . $user->getUserID();
```

File: src/PortlandLabs/ConcreteCmsTheme/User/EmptyAvatar.php [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+class EmptyAvatar implements AvatarInterface
+{
+
+    public function getPath()
+    {
+        return app(Repository::class)->get('concrete.icons.user_avatar.default');
+    }
+
+    public function output()
```
Comment (KorvinSzanto) [inline src/PortlandLabs/ConcreteCmsTheme/User/EmptyAvatar.php:17]:
```suggestion
    public function output(): string
```

File: src/PortlandLabs/ConcreteCmsTheme/User/EmptyAvatar.php [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+namespace PortlandLabs\ConcreteCmsTheme\User;
+
+use Concrete\Core\Config\Repository\Repository;
+use Concrete\Core\User\Avatar\AvatarInterface;
+use HtmlObject\Image;
+
+class EmptyAvatar implements AvatarInterface
+{
+
+    public function getPath()
```
Comment (KorvinSzanto) [inline src/PortlandLabs/ConcreteCmsTheme/User/EmptyAvatar.php:12]:
```suggestion
    public function getPath(): string
```

**Review by KorvinSzanto — CHANGES_REQUESTED** [review]
Few questions, needed changes, and suggestions

File: src/PortlandLabs/ConcreteCmsTheme/User/CommunityUser.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    ) {
+    }
+
+    public function getUserDisplayName(): string
+    {
+        $data = app(CommunityUserInspector::class)->getCommunityUserData($this->author);
+        $firstName = null;
+        $lastName = null;
+        if (isset($data['users']['id'])) {
+            foreach ($data['users']['custom_attributes']['custom_attributes'] as $attribute) {
```
[the remark this answers is not in the document]
Comment (aembler) [reply src/PortlandLabs/ConcreteCmsTheme/User/CommunityUser.php:24]: Yeah there's something weird in the API.

**Comment by aembler** [comment]
All changes should be accepted at this point.

**Pull request merged**